File: lib/utils/json_helpers.dart
/// Reusable encoding/decoding helpers for JSON maps and method channel results.
class JsonHelpers {
  /// Encodes an optional value to a JSON map if not null.
  ///
  /// [toJson] - Optional function to encode objects. If not provided, value is used directly.
  static void encode<T>(
    Map<String, dynamic> json,
    String key,
    T? value, [
    Map<String, dynamic> Function(T)? toJson,
  ]) {
    if (value != null) {
      json[key] = toJson != null ? toJson(value) : value;
    }
  }

  /// Encodes a list to a JSON map if not empty.
  static void encodeList<T>(
    Map<String, dynamic> json,
    String key,
    List<T> list,
    Map<String, dynamic> Function(T) toJson,
  ) {
    if (list.isNotEmpty) {
      json[key] = list.map(toJson).toList();
    }
  }

  /// Decodes an optional value from JSON maps or method channel results.
  ///
  /// [value] - The value to decode (typically `json[key]` or a method channel result).
  /// [fromJson] - Optional function to decode objects. If not provided, value is cast directly.
  static T? decode<T>(dynamic value, [T Function(Map)? fromJson]) {
    if (value == null) return null;
    if (fromJson != null) {
      return fromJson(value as Map);
    }
    return value as T?;
  }

  /// Decodes a list from JSON maps or method channel results, returning empty list if null.
  static List<T> decodeList<T>(List? value, T Function(Map) fromJson) {
    if (value == null) return const [];
    return value.map((e) => fromJson(e as Map)).toList();
  }

  /// Decodes an enum from its name string.
  ///
  /// [value] - The value to decode (will be cast to String). Can be null.
  /// [defaultValue] - Optional default value. If not provided, throws if [value] is null or doesn't match.
  /// Returns [defaultValue] if [value] is null or doesn't match (when provided), or throws otherwise.
  static T decodeEnum<T extends Enum>(
    dynamic value,
    List<T> values, [
    T? defaultValue,
  ]) {
    if (value == null) {
      if (defaultValue != null) return defaultValue;
      throw ArgumentError('Cannot decode null enum value');
    }
    final name = value as String;
    return values.firstWhere(
      (e) => e.name == name,
      orElse: () {
        if (defaultValue != null) return defaultValue;
        throw ArgumentError('No enum value found for "$name"');
      },
    );
  }

  /// Builds a concise `toString` output from non-null fields.
  static String formatToString(String type, Map<String, Object?> fields) {
    final buffer = StringBuffer('$type(');
    var first = true;
    fields.forEach((key, value) {
      if (value == null) return;
      if (!first) buffer.write(', ');
      first = false;
      buffer.write('$key: $value');
    });
    buffer.write(')');
    return buffer.toString();
  }
}

File: android/src/main/kotlin/co/quis/flutter_contacts/crud/models/JsonHelpers.kt
package co.quis.flutter_contacts.crud.models

/**
 * Reusable JSON encoding/decoding helpers.
 */
object JsonHelpers {
    /**
     * Encodes an optional value to JSON if not null.
     */
    fun <T> encodeOptional(
        json: MutableMap<String, Any?>,
        key: String,
        value: T?,
    ) {
        value?.let { json[key] = it }
    }

    /**
     * Encodes an optional object to JSON if not null.
     */
    fun <T> encodeOptionalObject(
        json: MutableMap<String, Any?>,
        key: String,
        value: T?,
        toJson: (T) -> Map<String, Any?>,
    ) {
        value?.let { json[key] = toJson(it) }
    }

    /**
     * Encodes a list to JSON if not empty.
     */
    fun <T> encodeList(
        json: MutableMap<String, Any?>,
        key: String,
        list: List<T>,
        toJson: (T) -> Map<String, Any?>,
    ) {
        if (list.isNotEmpty()) {
            json[key] = list.map(toJson)
        }
    }

    /**
     * Decodes an optional value from JSON.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T> decodeOptional(
        json: Map<String, Any?>,
        key: String,
    ): T? = json[key] as? T

    /**
     * Decodes an optional object from JSON.
     */
    fun <T> decodeOptionalObject(
        json: Map<String, Any?>,
        key: String,
        fromJson: (Map<String, Any?>) -> T,
    ): T? {
        @Suppress("UNCHECKED_CAST")
        val value = json[key] as? Map<String, Any?>
        return value?.let { fromJson(it) }
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> decodeRequiredObject(
        json: Map<String, Any?>,
        key: String,
        fromJson: (Map<String, Any?>) -> T,
    ): T {
        val value =
            json[key] as? Map<String, Any?>
                ?: throw IllegalArgumentException("Missing $key")
        return fromJson(value)
    }

    fun toStringKeyMap(map: Map<*, *>): Map<String, Any?> =
        map.entries
            .mapNotNull { (key, value) ->
                (key as? String)?.let { it to value }
            }.toMap()

    /**
     * Decodes a list from JSON, returning empty list if null or missing.
     */
    @Suppress("UNCHECKED_CAST")
    fun <T> decodeList(
        json: Map<String, Any?>,
        key: String,
        fromJson: (Map<String, Any?>) -> T,
    ): List<T> {
        val value = json[key] as? List<Map<String, Any?>>
        return value?.map { fromJson(it) } ?: emptyList()
    }

    /**
     * Decodes an optional timestamp (Number) from JSON as Long.
     */
    fun decodeOptionalTimestamp(
        json: Map<String, Any?>,
        key: String,
    ): Long? = (json[key] as? Number)?.toLong()
}
